### PR TITLE
Added custom status in user profile popover

### PR DESCRIPTION
**Pull request by jatinkksharma**

### CODE REVIEW

File: components/custom_status/custom_status_emoji.tsx [hunk 1]
```diff
@@ -16,30 +16,40 @@ interface ComponentProps {
...
+    let statusEmoji = (
         <RenderEmoji
             emoji={customStatus.emoji}
             size={emojiSize}
+            emojiStyle={emojiStyle}
         />
     );
+    let statusText = null;
+    if (showText) {
+        statusText = (
```
Comment (chetanyakan) [inline components/custom_status/custom_status_emoji.tsx]: Should we create another component to show both emoji and text??

File: components/custom_status/custom_status_emoji.tsx [hunk 2]
```diff
@@ -16,30 +16,40 @@ interface ComponentProps {
...
 }
 
-const CustomStatusEmoji = ({tooltipDirection = 'bottom', showTooltip = false, emojiSize = 16}: ComponentProps) => {
+const CustomStatusEmoji = ({tooltipDirection = 'bottom', showTooltip = false, emojiSize = 16, emojiStyle = {}, showText = false}: ComponentProps) => {
     const currentUser = useSelector((state: GlobalState) => getCurrentUser(state));
     if (!(currentUser && currentUser.props && currentUser.props.customStatus)) {
         return null;
     }
 
     const customStatus = JSON.parse(currentUser.props.customStatus);
```
Comment (chetanyakan) [inline components/custom_status/custom_status_emoji.tsx]: should we use the custom status selector??

File: components/profile_popover/profile_popover.jsx [hunk 3]
```diff
@@ -383,6 +386,41 @@ class ProfilePopover extends React.PureComponent {
...
 
+        const customStatus = this.props.customStatus;
+        if (customStatus) {
+            const customStatusEmoji = (
+                <span className='d-flex'>
+                    <CustomStatusEmoji
+                        size={14}
+                        showTooltip={false}
+                        emojiStyle={{
+                            marginRight: '4px',
```
Comment (chetanyakan) [inline components/profile_popover/profile_popover.jsx]:
```suggestion
                            marginRight: 4,
```

File: components/status_dropdown/index.jsx [hunk 4]
```diff
@@ -21,11 +22,13 @@ function mapStateToProps(state) {
...
     return {
-        currentUser,
-        profilePicture: Client4.getProfilePictureUrl(currentUser.id, currentUser.last_picture_update),
-        autoResetPref: get(state, Preferences.CATEGORY_AUTO_RESET_MANUAL_STATUS, currentUser.id, ''),
-        status: getStatusForUserId(state, currentUser.id),
+        userId,
+        profilePicture: Client4.getProfilePictureUrl(userId, currentUser.last_picture_update),
+        autoResetPref: get(state, Preferences.CATEGORY_AUTO_RESET_MANUAL_STATUS, userId, ''),
+        status: getStatusForUserId(state, userId),
+        customStatus: getCustomStatus(state),
```
Comment (chetanyakan) [inline components/status_dropdown/index.jsx]: Should we change this selector to pass an ID?
This way, we can get the status for all users.

We can default to the current user's status.

File: components/status_dropdown/status_dropdown.jsx [hunk 5]
```diff
@@ -141,25 +143,19 @@ export default class StatusDropdown extends React.PureComponent {
 
         let customStatusText = localizeMessage('status_dropdown.set_custom', 'Set a Custom Status');
         let customStatusEmoji = <EmojiIcon className={'custom-status-emoji'}/>;
-        const userProps = this.props.currentUser.props;
-        let isStatusSet = false;
-        if (userProps && userProps.customStatus) {
-            const customStatus = JSON.parse(this.props.currentUser.props.customStatus);
-            isStatusSet = customStatus.text || customStatus.emoji;
+        const customStatus = this.props.customStatus;
+        const isStatusSet = customStatus && (customStatus.text || customStatus.emoji);
```
Comment (chetanyakan) [inline components/status_dropdown/status_dropdown.jsx]: We should verify if we allow empty status texts.

File: selectors/views/custom_status.ts [hunk 6]
```diff
@@ -0,0 +1,17 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+import {getCurrentUser} from 'mattermost-redux/selectors/entities/users';
+
+import {GlobalState} from 'types/store';
+
+export function getCustomStatus(state: GlobalState) {
```
Comment (chetanyakan) [inline selectors/views/custom_status.ts]: Same here, alow passing a `userID`.

**Review by chetanyakan — CHANGES_REQUESTED** [review]

**Comment by chetanyakan** [comment]
1. UX Clarification
2. Slash command - investigation

File: components/channel_header/channel_header.js [hunk 7]
```diff
@@ -417,19 +421,35 @@ class ChannelHeader extends React.PureComponent {
...
             );
 
+            dmHeaderCustomStatus = this.props.isCustomStatusEnabled && (
+                <>
+                    <CustomStatusEmoji
+                        userID={this.props.dmUser.id}
+                        emojiSize={15}
+                        emojiStyle={{
+                            verticalAlign: 'top',
+                            margin: '0px 4px',
```
Comment (chetanyakan) [inline components/channel_header/channel_header.js]: No need to specify dimensions for `0`

```suggestion
                            margin: '0 4px',
```

File: components/custom_status/custom_status_emoji.tsx [hunk 8]
```diff
@@ -4,31 +4,40 @@ import React from 'react';
...
+    const customStatusEnabled = useSelector((state: GlobalState) => {
+        return isCustomStatusEnabled(state);
+    });
+    const customStatus = useSelector((state: GlobalState) => {
+        return getCustomStatus(state, userID);
+    });
+    if (!(customStatus && customStatus.emoji)) {
+        return null;
+    }
+    if (!customStatusEnabled) {
```
Comment (chetanyakan) [inline components/custom_status/custom_status_emoji.tsx]: nit - Merge these checks

```suggestion
    if (!customStatusEnabled || !customStatus || !customStatus.emoji) {
```

File: components/custom_status/custom_status_emoji.tsx [hunk 9]
```diff
@@ -4,31 +4,40 @@ import React from 'react';
...
 }
 
-const CustomStatusEmoji = ({tooltipDirection = 'bottom', showTooltip = false, emojiSize = 16}: ComponentProps) => {
-    const currentUser = useSelector((state: GlobalState) => getCurrentUser(state));
-    if (!(currentUser && currentUser.props && currentUser.props.customStatus)) {
+const CustomStatusEmoji = (props: ComponentProps) => {
+    const {emojiSize, emojiStyle, showTooltip, tooltipDirection, userID} = props;
+    const customStatusEnabled = useSelector((state: GlobalState) => {
+        return isCustomStatusEnabled(state);
+    });
```
Comment (chetanyakan) [inline components/custom_status/custom_status_emoji.tsx:26]: If `isCustomStatusEnabled` is an arrow function, would this work?

```suggestion
    const customStatusEnabled = useSelector(isCustomStatusEnabled);
```

If so, make similar changes elsewhere.

File: components/custom_status/custom_status_emoji.tsx [hunk 10]
```diff
@@ -39,22 +48,36 @@ const CustomStatusEmoji = ({tooltipDirection = 'bottom', showTooltip = false, em
...
                 overlay={
-                    <Tooltip id='custom-status'>
+                    <Tooltip id='custom-status-tooltip'>
                         <div className='custom-status'>
-                            {statusEmoji}
+                            <RenderEmoji
+                                emoji={customStatus.emoji}
+                                size={14}
+                            />
+                            {' '}
```
Comment (chetanyakan) [inline components/custom_status/custom_status_emoji.tsx:55]: Check if we should use the `Nbsp` component instead.

```ts
import Nbsp from 'components/html_entities/nbsp';
```

File: components/custom_status/custom_status_modal.tsx [hunk 11]
```diff
@@ -24,6 +24,8 @@ type Props = {
     onHide: () => void;
 };
 
+const EMOJI_PICKER_WIDTH_OFFSET = 308;
```
Comment (chetanyakan) [inline components/custom_status/custom_status_modal.tsx:27]: Should we move this with the rest of the constants?

File: components/custom_status/custom_status_modal.tsx [hunk 12]
```diff
@@ -119,6 +121,20 @@ const CustomStatusModal: React.FC<Props> = (props: Props) => {
...
+            const anyTarget: any = target;
+            rightOffset = window.innerWidth - anyTarget.getBoundingClientRect().left - EMOJI_PICKER_WIDTH_OFFSET;
+            if (rightOffset < 0) {
+                rightOffset = Constants.DEFAULT_EMOJI_PICKER_RIGHT_OFFSET;
+            }
+        }
+
+        return rightOffset;
+    };
+
```
Comment (chetanyakan) [inline components/custom_status/custom_status_modal.tsx:137]: Should we move this to `utils`?

File: components/post_view/post_header/index.ts [hunk 13]
```diff
@@ -24,14 +31,26 @@ function mapStateToProps(state: GlobalState, ownProps: Props) {
...
+    const customStatus = user ? getCustomStatus(state, user.id) : {};
     const isBot = Boolean(user && user.is_bot);
 
     return {
         enablePostUsernameOverride,
         isBot,
         overwriteIcon,
         isGuest: Boolean(user && isGuest(user)),
+        customStatus,
+        userID: currentUser.id,
```
Comment (chetanyakan) [inline components/post_view/post_header/index.ts]: Rename to `currentUserID`

```suggestion
        currentUserID: currentUser.id,
```

File: components/post_view/post_header/post_header.tsx [hunk 14]
```diff
@@ -84,9 +88,33 @@ export type Props = {
...
+
+    /**
+     * Custom status of the user
+     */
+    customStatus: CustomStatus;
+
+    /**
+     * User id of logged in user.
+     */
+    userID: string;
```
Comment (chetanyakan) [inline components/post_view/post_header/post_header.tsx]: Same here. Rename prop

File: components/post_view/post_header/index.ts [hunk 15]
```diff
@@ -24,14 +31,26 @@ function mapStateToProps(state: GlobalState, ownProps: Props) {
     }
 
     const user = getUser(state, ownProps.post.user_id);
+    const currentUser = getCurrentUser(state);
+    const customStatus = user ? getCustomStatus(state, user.id) : {};
```
Comment (chetanyakan) [inline components/post_view/post_header/index.ts:35]: Should this user be `ownProps.post.user_id`?
```suggestion
    const customStatus = getCustomStatus(state, ownProps.post.user_id) || {};
```

File: components/post_view/post_header/post_header.tsx [hunk 16]
```diff
@@ -167,12 +196,43 @@ export default class PostHeader extends React.PureComponent<Props> {
...
+        const isCustomStatusSet = userCustomStatus && userCustomStatus.emoji;
+        const isCurrentUser = this.props.post.user_id && this.props.post.user_id === this.props.userID
+        if (!isSystemMessage && isCustomStatusSet) {
+            customStatus = (
+                <CustomStatusEmoji
+                    userID={this.props.post.user_id}
+                    showTooltip={true}
+                    emojiSize={14}
+                    emojiStyle={{
+                        margin: '4px 0px 0px 4px',
```
Comment (chetanyakan) [inline components/post_view/post_header/post_header.tsx]: nit - 
```suggestion
                        margin: '4px 0 0 4px',
```

Also, check if shorthand `0 4px` works

File: components/profile_popover/profile_popover.jsx [hunk 17]
```diff
@@ -383,6 +403,64 @@ class ProfilePopover extends React.PureComponent {
...
+                        </div>
+                    </div>
+                );
+            } else if (this.props.user.id === this.props.currentUserId) {
+                customStatusContent = (
+                    <div>
+                        <a
+                            href='#'
+                            onClick={this.showCustomStatusModal}
+                        >
```
Comment (chetanyakan) [inline components/profile_popover/profile_popover.jsx]: 1. Check if we should use something other than the `<a>` tag.
2. Add appropriate `aria` properties

**Review by chetanyakan — CHANGES_REQUESTED** [review]

File: components/custom_status/custom_status_modal.tsx (hunk 11 above)
Comment (jatinkksharma) [reply components/custom_status/custom_status_modal.tsx:27]: Doesn't make sense to move this in global constants, as the constant is only related to this component and will not be reused.

File: components/custom_status/custom_status_emoji.tsx (hunk 9 above)
Comment (jatinkksharma) [reply components/custom_status/custom_status_emoji.tsx:26]: `isCustomStatusEnabled` is not an arrow function.

File: components/post_view/post_header/index.ts (hunk 15 above)
Comment (jatinkksharma) [reply components/post_view/post_header/index.ts:35]: `ownProps.post.user_id` is same as `user.id`.

File: components/custom_status/custom_status_emoji.tsx (hunk 10 above)
Comment (jatinkksharma) [reply components/custom_status/custom_status_emoji.tsx:55]: The return type of Nbsp is ReactNode and the return type of custom_status_emoji is JSx component. So it gives typescript error to use Nbsp in custom_status_emoji  component.

**Review by chetanyakan — APPROVED** [review]

**Pull request merged**